kernel crashes with small CONFIG_TIMESLICE_SIZE

**Describe the bug**
For some time now we've received reports that two tests are failing sporadically in CI for mysterious reasons, #13813 and #19701. As it turns out, what these tests have in common is that in their prj.conf, they both set `CONFIG_TIMESLICE_SIZE=1`.

There seems to be some kind of race condition in the kernel which is made worse by having timeslicing set to this. Now that almost all of our platforms enable icount in the QEMU configuration, this now doesn't happen randomly as all tests run in a deterministic fashion and it's harder to get unlucky.

Looking through test configurations, we do not have any dedicated tests for timeslicing. Only three tests change the default value of CONFIG_TIMESLICE_SIZE=0, which means unless I'm missing something, only these tests are exercising the timeslice feature, and all of them are actually testing something else:

tests/kernel/mem_slab/mslab_threadsafe/prj.conf sets it to 1
tests/kernel/mem_pool/mem_pool_threadsafe/prj.conf sets it to 1
tests/kernel/mem_protect/syscalls/prj.conf sets it to 10 (for the syscall torture test)

**To Reproduce**

To best show this issue, we need to default the timeslice size to 1 and disable icount:

```
diff --git a/cmake/emu/qemu.cmake b/cmake/emu/qemu.cmake
index 26e7622dee..a05ca93cd3 100644
--- a/cmake/emu/qemu.cmake
+++ b/cmake/emu/qemu.cmake
@@ -60,12 +60,6 @@ endif()
 # Connect monitor to the console chardev.
 list(APPEND QEMU_FLAGS -mon chardev=con,mode=readline)

-if(CONFIG_QEMU_ICOUNT)
-  list(APPEND QEMU_FLAGS
-         -icount shift=${CONFIG_QEMU_ICOUNT_SHIFT},align=off,sleep=off
-         -rtc clock=vm)
-endif()
-
 # Add a BT serial device when building for bluetooth, unless the
 # application explicitly opts out with NO_QEMU_SERIAL_BT_SERVER.
 if(CONFIG_BT)
diff --git a/kernel/Kconfig b/kernel/Kconfig
index dda626e136..198f646744 100644
--- a/kernel/Kconfig
+++ b/kernel/Kconfig
@@ -399,7 +399,7 @@ config TIMESLICING

 config TIMESLICE_SIZE
        int "Time slice size (in ms)"
-       default 0
+       default 1
        range 0 2147483647
        depends on TIMESLICING
        help
```

Then run sanitycheck. In a typical run I would see 

There were 43 total failures when testing `sanitycheck -T tests/kernel` which seems abnormal even with icount disabled. I looked through the logs:

- There were a number of expected timing ordering failures that we've always known about (and why we brought in icount), maybe 5-10
- Several tests failed with no error in the output, as if the output was truncated or the emulator exited suddenly
- I also saw lots of crashes. The majority of the failures were with qemu_cortex_m0 or qemu_cortex_r5, here are some examples:

qemu_cortex_r5            tests/kernel/mbox/mbox_api/kernel.mailbox.api
```
START - test_mbox_async_put_to_waiting_get
E: ***** HARD FAULT *****
E: r0/a1:  0xfffffffd  r1/a2:  0xe000ed00  r2/a3:  0xffffffff
E: r3/a4:  0xffffffff r12/ip:  0x00000000 r14/lr:  0x00003f87
E:  xpsr:  0x81000000
E: Faulting instruction address (r15/pc): 0x00005dec
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x200005a0 (unknown)
E: Halting system
```

qemu_cortex_m0            tests/kernel/sched/schedule_api/kernel.scheduler.multiq
```
START - test_lock_preemptible
E: ***** HARD FAULT *****
E: r0/a1:  0xfffffffd  r1/a2:  0xe000ed00  r2/a3:  0xffffffff
E: r3/a4:  0xffffffff r12/ip:  0x00000000 r14/lr:  0x00003ee3
E:  xpsr:  0x81000000
E: Faulting instruction address (r15/pc): 0x00005a84
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20000758 (unknown)
E: Halting system
```

qemu_cortex_m0            tests/kernel/sched/schedule_api/kernel.scheduler
```
START - test_lock_preemptible
E: ***** HARD FAULT *****
E: r0/a1:  0xfffffffd  r1/a2:  0xe000ed00  r2/a3:  0xffffffff
E: r3/a4:  0xffffffff r12/ip:  0x00000000 r14/lr:  0x00003ee3
E:  xpsr:  0x81000000
E: Faulting instruction address (r15/pc): 0x00005aa4
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20000758 (unknown)
E: Halting system
```

qemu_cortex_r5            tests/kernel/mem_protect/sys_sem/kernel.memory_protection.sys_sem.nouser
```
START - test_sem_take_multiple
ASSERTION FAIL [thread->base.pended_on] @ WEST_TOPDIR/zephyr/kernel/sched.c:475
@ WEST_TOPDIR/zephyr/lib/os/assert.c:45
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x000060b0 (unknown)
E: Halting system
```

qemu_cortex_r5            tests/kernel/mem_pool/mem_pool_threadsafe/kernel.memory_pool.threadsafe
```
START - test_mpool_threadsafe
E: ***** PREFETCH ABORT *****
E: Permission Fault @ 0xffb41092
E: r0/a1:  0x00004c58  r1/a2:  0x00004ec0  r2/a3:  0x00004ec0
E: r3/a4:  0x00000000 r12/ip:  0x00000002 r14/lr:  0x00003098
E:  xpsr:  0x600001ff
E: Faulting instruction address (r15/pc): 0xffb41092
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x00004ec0 (main)
E: Halting system
```

caught this failing a timeout assertion on x86_64:

```
Booting from ROM..*** Booting Zephyr OS build zephyr-v2.3.0-1198-g849ff583968e  ***
Running test suite mpool_threadsafe
===================================================================
START - test_mpool_threadsafe
ASSERTION FAIL [!sys_dnode_is_linked(&to->node)] @ WEST_TOPDIR/zephyr/kernel/timeout.c:103
	
E: RAX: 0x0000000000000004 RBX: 0x0000000000109390 RCX: 0x0000000000000001 RDX: 0x0000000000000046
E: RSI: 0x0000000000000067 RDI: 0x000000000010847b RBP: 0x0000000000000015 RSP: 0x0000000000111088
E: Page fault at address 0xb9 (error code 0x11)
E:  R8: 0x0000000000000001  R9: 0x0000000000000020 R10: 0x0000000000000202 R11: 0x0000000000111098
E: Access violation: supervisor thread not allowed to execute
E: R12: 0xfffffffffffffffe R13: 0x00000000001034c3 R14: 0x0000000000000016 R15: 0x0000000000000010
E: PTE: 0x0000000000000000 -> 0x0000000000000000: RW XD 
E: RAX: 0x0000000000000001 RBX: 0x0000000000111050 RCX: 0x0000000000000001 RDX: 0x0000000000000046
E: RSI: 0x0000000000000000 RDI: 0x000000000010bc64 RBP: 0x00000000001110d0 RSP: 0x0000000000111048
E:  R8: 0x000000000010bb30  R9: 0x0000000000000001 R10: 0x0000000000000202 R11: 0x0000000000109da0
E: RSP: 0x0000000000111088 RFLAGS: 0x0000000000000006 CS: 0x0018 CR3: 0x0000000000141000
E: R12: 0x0000000000000000 R13: 0x0000000000000001 R14: 0x000000000010bb10 R15: 0x000000000013c1c0
E: RSP: 0x0000000000111048 RFLAGS: 0x0000000000000247 CS: 0x0018 CR3: 0x0000000000141000
E: RIP: 0x00000000000000b9
E: RIP: 0x0000000000100dde
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 1
E: Attempt to resume un-suspended thread object
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x0000000000109360 (unknown)
E: Current thread: 0x0000000000109360 (unknown)
E: Halting system

```

**Expected behavior**
No failures like this with timeslicing turned on.

**Impact**
This appears to be a showstopper for anyone who uses time slicing, and depending on where the root cause is, others too.

Found something interesting.

I ran some sanitycheck with this revision of https://github.com/andrewboie/zephyr/tree/this-should-not-happen-27533  which is a version of https://github.com/zephyrproject-rtos/zephyr/pull/27995 plus this patch applied:

```
diff --git a/kernel/Kconfig b/kernel/Kconfig
index 3e72f4acfd..d80d79e4d1 100644
--- a/kernel/Kconfig
+++ b/kernel/Kconfig
@@ -400,7 +400,7 @@ config TIMESLICING

 config TIMESLICE_SIZE
        int "Time slice size (in ms)"
-       default 0
+       default 5
        range 0 2147483647
        depends on TIMESLICING
        help
```

The test tests/kernel/mem_protect/sys_sem fails like so. Thankfully we have icount and it does it every time:

```
START - test_sem_take_multiple
ASSERTION FAIL [0] @ WEST_TOPDIR/zephyr/kernel/thread_abort.c:47
        should never get here
E: r0/a1:  0x00000004  r1/a2:  0x0000002f  r2/a3:  0x00000000
E: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x0000771b
E:  xpsr:  0x61000000
E: Faulting instruction address (r15/pc): 0x0000b406
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
E: Current thread: 0x200005c0 (unknown)
E: Halting system
qemu-system-arm: terminating on signal 2
ninja: build stopped: interrupted by user.
```

The assertion in question is here:

```
FUNC_NORETURN void z_self_abort(void)
{
	/* Self-aborting threads don't clean themselves up, we
	 * have the idle thread for the current CPU do it.
	 */
	int key;
	struct _cpu *cpu;

	/* Lock local IRQs to prevent us from migrating to another CPU
	 * while we set this up
	 */
	key = arch_irq_lock();
	cpu = _current_cpu;
	cpu->pending_abort = _current;

	LOG_DBG("%p self-aborting, handle on idle thread %p",
		_current, cpu->idle_thread);

	k_thread_suspend(_current);
	z_swap_irqlock(key);
	__ASSERT(false, "should never get here");
	CODE_UNREACHABLE;
}
```

We're in here because the thread exited. It is calling `k_thread_suspend()` on itself (with IRQs locked so execution continues) and then z_swap_irqlock(), and we should never get back from that because we have suspended ourselves; the idea is the next context switch will be to idle to clean up. But instead we are somehow getting scheduled again and the assertion fails.

The thread is `sem_tid2` in the test case, which runs this function:

```
void sem_take_multiple_high_prio_helper(void *p1, void *p2, void *p3)
{
	int32_t ret_value;

	ret_value = sys_sem_take(&high_prio_sem, K_FOREVER);
	zassert_true(ret_value == 0, "sys_sem_take failed");

	ret_value = sys_sem_take(&multiple_thread_sem, K_FOREVER);
	zassert_true(ret_value == 0, "sys_sem_take failed");

	sys_sem_give(&high_prio_sem);
}
```

The documentation says:

```
**
 * @brief Suspend a thread.
 *
 * This routine prevents the kernel scheduler from making @a thread
 * the current thread. All other internal operations on @a thread are
 * still performed; for example, kernel objects it is waiting on are
 * still handed to it.  Note that any existing timeouts
 * (e.g. k_sleep(), or a timeout argument to k_sem_take() et. al.)
 * will be canceled.  On resume, the thread will begin running
 * immediately and return from the blocked call.
 *
 * If @a thread is already suspended, the routine has no effect.
```

The test case is a set of threads giving and taking semaphores on each other. Since the thread self-suspended, it's not waiting on anything, so "kernel objects it is waiting on are still handed to it" shouldn't apply. Yet is waking up anyway. It's not clear why it's being woken up like this.

The whole thing is very timing sensitive, adding printks() or enabling logs causes it to go away.

Apologies if this is a wild goose chase and my `z_self_abort()` is unsound.

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -630,12 +630,17 @@ ALWAYS_INLINE struct k_thread *z_find_first_thread_to_unpend(_wait_q_t *wait_q,
 	return ret;
 }
 
+static inline void unpend_thread_no_timeout(struct k_thread *thread)
+{
+	_priq_wait_remove(&pended_on(thread)->waitq, thread);
+	z_mark_thread_as_not_pending(thread);
+	thread->base.pended_on = NULL;
+}
+
 ALWAYS_INLINE void z_unpend_thread_no_timeout(struct k_thread *thread)
 {
 	LOCKED(&sched_spinlock) {
-		_priq_wait_remove(&pended_on(thread)->waitq, thread);
-		z_mark_thread_as_not_pending(thread);
-		thread->base.pended_on = NULL;
+		unpend_thread_no_timeout(thread);
 	}
 }
 
@@ -643,15 +648,17 @@ ALWAYS_INLINE void z_unpend_thread_no_timeout(struct k_thread *thread)
 /* Timeout handler for *_thread_timeout() APIs */
 void z_thread_timeout(struct _timeout *timeout)
 {
-	struct k_thread *thread = CONTAINER_OF(timeout,
-					       struct k_thread, base.timeout);
+	LOCKED(&sched_spinlock) {
+		struct k_thread *thread = CONTAINER_OF(timeout,
+						struct k_thread, base.timeout);
 
-	if (thread->base.pended_on != NULL) {
-		z_unpend_thread_no_timeout(thread);
+		if (thread->base.pended_on != NULL) {
+			unpend_thread_no_timeout(thread);
+		}
+		z_mark_thread_as_started(thread);
+		z_mark_thread_as_not_suspended(thread);
+		ready_thread(thread);
 	}
-	z_mark_thread_as_started(thread);
-	z_mark_thread_as_not_suspended(thread);
-	z_ready_thread(thread);
 }
 #endif
 

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -303,9 +303,9 @@ void k_sched_time_slice_set(int32_t slice, int prio)
 static inline int sliceable(struct k_thread *thread)
 {
 	return is_preempt(thread)
+		&& !z_is_thread_prevented_from_running(thread)
 		&& !z_is_prio_higher(thread->base.prio, slice_max_prio)
-		&& !z_is_idle_thread_object(thread)
-		&& !z_is_thread_timeout_active(thread);
+		&& !z_is_idle_thread_object(thread);
 }
 
 /* Called out of each timer interrupt */

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -57,6 +57,8 @@ struct z_kernel _kernel;
 
 static struct k_spinlock sched_spinlock;
 
+static void update_cache(int);
+
 #define LOCKED(lck) for (k_spinlock_key_t __i = {},			\
 					  __key = k_spin_lock(lck);	\
 			!__i.key;					\
@@ -264,6 +266,16 @@ static ALWAYS_INLINE struct k_thread *next_up(void)
 #endif
 }
 
+static void move_thread_to_end_of_prio_q(struct k_thread *thread)
+{
+	if (z_is_thread_queued(thread)) {
+		_priq_run_remove(&_kernel.ready_q.runq, thread);
+	}
+	_priq_run_add(&_kernel.ready_q.runq, thread);
+	z_mark_thread_as_queued(thread);
+	update_cache(thread == _current);
+}
+
 #ifdef CONFIG_TIMESLICING
 
 static int slice_time;
@@ -311,9 +323,19 @@ static inline int sliceable(struct k_thread *thread)
 /* Called out of each timer interrupt */
 void z_time_slice(int ticks)
 {
+	/* Hold sched_spinlock, so that activity on another CPU
+	 * (like a call to k_thread_abort() at just the wrong time)
+	 * won't affect the correctness of the decisions made here.
+	 * Also prevents any nested interrupts from changing
+	 * thread state to avoid similar issues, since this would
+	 * normally run with IRQs enabled.
+	 */
+	k_spinlock_key_t key = k_spin_lock(&sched_spinlock);
+
 #ifdef CONFIG_SWAP_NONATOMIC
 	if (pending_current == _current) {
 		z_reset_time_slice();
+		k_spin_unlock(&sched_spinlock, key);
 		return;
 	}
 	pending_current = NULL;
@@ -321,7 +343,7 @@ void z_time_slice(int ticks)
 
 	if (slice_time && sliceable(_current)) {
 		if (ticks >= _current_cpu->slice_ticks) {
-			z_move_thread_to_end_of_prio_q(_current);
+			move_thread_to_end_of_prio_q(_current);
 			z_reset_time_slice();
 		} else {
 			_current_cpu->slice_ticks -= ticks;
@@ -329,6 +351,7 @@ void z_time_slice(int ticks)
 	} else {
 		_current_cpu->slice_ticks = 0;
 	}
+	k_spin_unlock(&sched_spinlock, key);
 }
 #endif
 
@@ -401,12 +424,7 @@ void z_ready_thread(struct k_thread *thread)
 void z_move_thread_to_end_of_prio_q(struct k_thread *thread)
 {
 	LOCKED(&sched_spinlock) {
-		if (z_is_thread_queued(thread)) {
-			_priq_run_remove(&_kernel.ready_q.runq, thread);
-		}
-		_priq_run_add(&_kernel.ready_q.runq, thread);
-		z_mark_thread_as_queued(thread);
-		update_cache(thread == _current);
+		move_thread_to_end_of_prio_q(thread);
 	}
 }
 

--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -1221,21 +1221,16 @@ static int32_t z_tick_sleep(int32_t ticks)
 
 	expected_wakeup_time = ticks + z_tick_get_32();
 
-	/* Spinlock purely for local interrupt locking to prevent us
-	 * from being interrupted while _current is in an intermediate
-	 * state.  Should unify this implementation with pend().
-	 */
-	struct k_spinlock local_lock = {};
-	k_spinlock_key_t key = k_spin_lock(&local_lock);
+	k_spinlock_key_t key = k_spin_lock(&sched_spinlock);
 
 #if defined(CONFIG_TIMESLICING) && defined(CONFIG_SWAP_NONATOMIC)
 	pending_current = _current;
 #endif
-	z_remove_thread_from_ready_q(_current);
+	unready_thread(_current);
 	z_add_thread_timeout(_current, timeout);
 	z_mark_thread_as_suspended(_current);
 
-	(void)z_swap(&local_lock, key);
+	(void)z_swap(&sched_spinlock, key);
 
 	__ASSERT(!z_is_thread_state_set(_current, _THREAD_SUSPENDED), "");
 

